src/_pytest/faulthandler.py

```diff
@@ -17,70 +17,92 @@ def pytest_addoption(parser):
 def pytest_configure(config):
     import faulthandler
 
-    # avoid trying to dup sys.stderr if faulthandler is already enabled
-    if faulthandler.is_enabled():
-        return
-
-    stderr_fd_copy = os.dup(_get_stderr_fileno())
-    config.fault_handler_stderr = os.fdopen(stderr_fd_copy, "w")
-    faulthandler.enable(file=config.fault_handler_stderr)
-
-
-def _get_stderr_fileno():
-    try:
-        return sys.stderr.fileno()
-    except (AttributeError, io.UnsupportedOperation):
-        # python-xdist monkeypatches sys.stderr with an object that is not an actual file.
-        # https://docs.python.org/3/library/faulthandler.html#issue-with-file-descriptors
-        # This is potentially dangerous, but the best we can do.
-        return sys.__stderr__.fileno()
+    if not faulthandler.is_enabled():
+        # faulthhandler is not enabled, so install plugin that does the actual work
+        # of enabling faulthandler before each test executes.
+        config.pluginmanager.register(FaultHandlerHooks(), "faulthandler-hooks")
+    else:
+        from _pytest.warnings import _issue_warning_captured
+
+        # Do not handle dumping to stderr if faulthandler is already enabled, so warn
+        # users that the option is being ignored.
+        timeout = FaultHandlerHooks.get_timeout_config_value(config)
+        if timeout > 0:
+            _issue_warning_captured(
+                pytest.PytestConfigWarning(
+                    "faulthandler module enabled before pytest configuration step, "
+                    "'faulthandler_timeout' option ignored"
+                ),
+                config.hook,
+                stacklevel=2,
+            )
+
+
+class FaultHandlerHooks:
+    """Implements hooks that will actually install fault handler before tests execute,
+    as well as correctly handle pdb and internal errors."""
+
+    def pytest_configure(self, config):
+        import faulthandler
 
+        stderr_fd_copy = os.dup(self._get_stderr_fileno())
+        config.fault_handler_stderr = os.fdopen(stderr_fd_copy, "w")
+        faulthandler.enable(file=config.fault_handler_stderr)
 
-def pytest_unconfigure(config):
-    import faulthandler
+    def pytest_unconfigure(self, config):
+        import faulthandler
 
-    faulthandler.disable()
-    # close our dup file installed during pytest_configure
-    f = getattr(config, "fault_handler_stderr", None)
```

Comment:
(Haven't read the discussion, just something I noticed)

@nicoddemus I've heard a rumor that it's possible for `pytest_unconfigure` to be called without `pytest_configure` having been called:

https://github.com/pytest-dev/pytest/blob/b7ad4c2bed3906e1a12d9bb1829328b4045a72de/src/_pytest/debugging.py#L61-L62

Dunno if it's relevant here, but maybe that's why it was written in this defensive style. cc @blueyed e61e81a7b53496ea53f1f9958a2be5468190b8ef.

Reply:
Thanks for pointing that out @bluetech.

We definitely should fix pytest though so `pytest_unconfigure` is only called for `pytest_configure`, although this might need changes in `pluggy`: otherwise we make things very tricky for plugin authors.

Cc @goodboy

Reply:
Not sure I follow @nicoddemus.
The calling of each is determined by an internal `pytest` protocol no?

FD I've only briefed this PR.

Reply:
IIRC it might happen if a plugin raises UsageError, that then `pytest_unconfigure` gets called although `pytest_configure` was not.  (there is a test for this somewhere)
Yes, that behavior is related to pytest, not pluggy I assume.

Reply:
To exemplify:

Suppose we have plugins `A` and `B`. pytest will call `pytest_configure` and `pytest_unconfigure` using something like this:

```python
hook.pytest_configure(config=config)
```

Now suppose `A.pytest_configure` raises `UsageError`. From the call above currently has no way to know that `B.pytest_configure` never got called due to the error being raised.

Later when pytest tries to call:

```python
hook.pytest_unconfigure(config=config)
```

There's now way for pytest to know that it should not call `B.pytest_unconfigure` (because `B.pytest_configure` was never called).

So unless I'm missing something, we need some support in pluggy to be able to handle this in pytest?